src/server/command/worldedit/pattern.zig

```diff
@@ -0,0 +1,58 @@
+const std = @import("std");
+
+const main = @import("root");
+const AliasTable = main.utils.AliasTable;
+const Block = main.blocks.Block;
+const ListUnmanaged = main.ListUnmanaged;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+
+blocks: AliasTable(Entry),
+
+const Entry = struct {
+	block: Block,
+	chance: f32,
+};
+
+pub fn initFromString(allocator: NeverFailingAllocator, source: []const u8) !@This() {
+	var specifiers = std.mem.splitScalar(u8, source, ',');
+	var totalWeight: usize = 0;
+
+	var weightedEntries: ListUnmanaged(struct {block: Block, weight: usize}) = .{};
+	defer weightedEntries.deinit(main.stackAllocator);
+
+	while(specifiers.next()) |specifier| {
+		var iterator = std.mem.splitScalar(u8, specifier, '%');
+
+		const first = iterator.next();
+		const second = iterator.next();
+
+		if(iterator.peek() != null) {
+			return error.PatternSyntaxError;
+		}
+		if(second) |blockId| {
+			if(first) |weight| {
+				const block = main.blocks.parseBlock(blockId);
+				const blockWeight = try std.fmt.parseInt(usize, weight, 10);
+				totalWeight += blockWeight;
+				weightedEntries.append(main.stackAllocator, .{.block = block, .weight = blockWeight});
+				continue;
+			} else return error.PatternSyntaxError;
+		}
+		if(first) |blockId| {
+			const block = main.blocks.parseBlock(blockId);
+			totalWeight += 1;
+			weightedEntries.append(main.stackAllocator, .{.block = block, .weight = 1});
+		} else return error.PatternSyntaxError;
+	}
+
+	const entries = allocator.alloc(Entry, weightedEntries.items.len);
```

Comment:
I think this is leaked. I don't think the AliasTable frees the pointers you give it.
You may need to manually clean it in the deinit function, or add a way to transfer ownership of the pointer to the AliasTable.

Reply:
Yes, this is definitely leaked when AliasTable is initialized with `.init()`, I checked that and forgot to add it to `deinit()`